Rustdoc shows broken anonymous lifetimes on re-exported signatures

```toml
# Cargo.toml

[package]
name = "repro"
version = "0.0.0"
edition = "2021"
publish = false

[dependencies]
foo = { path = "foo" }
```

```rust
// src/lib.rs

pub use foo::repro;
```

```toml
# foo/Cargo.toml

[package]
name = "foo"
version = "0.0.0"
edition = "2021"
publish = false
```

```rust
// foo/src/lib.rs

pub fn repro<F>()
where
    F: Fn(&str),
{
    unimplemented!()
}
```

The correct behavior is for both crates' docs to render `repro` as:

<img src="https://user-images.githubusercontent.com/1940490/176577639-bd3508ac-5d84-463f-8987-c57af472dbc0.png" height="100">

However, recent versions of nightly rustdoc render `foo::repro` as the above, while rendering `repro::repro` incorrectly as:

<img src="https://user-images.githubusercontent.com/1940490/176577739-562a6382-9592-4e98-8080-912fe95be5a7.png" height="100">

I believe this is a bug because that is not valid syntax in Rust.

```console
error[E0637]: `'_` cannot be used here
 --> src/lib.rs:3:12
  |
3 |     F: for<'_> Fn(&str),
  |            ^^ `'_` is a reserved lifetime name
```

--- a/src/librustdoc/clean/mod.rs
+++ b/src/librustdoc/clean/mod.rs
@@ -167,7 +167,7 @@ fn clean_poly_trait_ref_with_bindings<'tcx>(
         .collect_referenced_late_bound_regions(&poly_trait_ref)
         .into_iter()
         .filter_map(|br| match br {
-            ty::BrNamed(_, name) => Some(GenericParamDef {
+            ty::BrNamed(_, name) if name != kw::UnderscoreLifetime => Some(GenericParamDef {
                 name,
                 kind: GenericParamDefKind::Lifetime { outlives: vec![] },
             }),

--- a/src/test/rustdoc/auxiliary/issue-98697-reexport-with-anonymous-lifetime.rs
+++ b/src/test/rustdoc/auxiliary/issue-98697-reexport-with-anonymous-lifetime.rs
@@ -1,0 +1,9 @@
+/// When reexporting this function, make sure the anonymous lifetimes are not rendered.
+///
+/// https://github.com/rust-lang/rust/issues/98697
+pub fn repro<F>()
+where
+    F: Fn(&str),
+{
+    unimplemented!()
+}

--- a/src/test/rustdoc/issue-98697.rs
+++ b/src/test/rustdoc/issue-98697.rs
@@ -1,0 +1,13 @@
+// aux-build:issue-98697-reexport-with-anonymous-lifetime.rs
+// ignore-cross-compile
+
+// When reexporting a function with a HRTB with anonymous lifetimes,
+// make sure the anonymous lifetimes are not rendered.
+//
+// https://github.com/rust-lang/rust/issues/98697
+
+extern crate issue_98697_reexport_with_anonymous_lifetime;
+
+// @has issue_98697/fn.repro.html '//pre[@class="rust fn"]/code' 'fn repro<F>() where F: Fn(&str)'
+// @!has issue_98697/fn.repro.html '//pre[@class="rust fn"]/code' 'for<'
+pub use issue_98697_reexport_with_anonymous_lifetime::repro;

